src/cursor.c

```diff
@@ -14,26 +14,100 @@
 #include "common/scene-helpers.h"
 #include "common/zfree.h"
 
-static const struct cursor_alias {
-	const char *name, *alias;
-} cursor_aliases[] = {
-	{ "default",    "left_ptr" },
-	{ "text",       "xterm" },
-	{ "grab",       "grabbing" },
-	{ "pointer",    "hand1" },
-	{ "wait",       "watch" },
-	{ "all-scroll", "grabbing" },
-	/* Resize edges */
-	{ "nw-resize",  "top_left_corner" },
-	{ "n-resize",   "top_side" },
-	{ "ne-resize",  "top_right_corner" },
-	{ "e-resize",   "right_side" },
-	{ "se-resize",  "bottom_right_corner" },
-	{ "s-resize",   "bottom_side" },
-	{ "sw-resize",  "bottom_left_corner" },
-	{ "w-resize",   "left_side" },
+static const char **cursor_names = NULL;
+
+/* Usual cursor names */
+static const char *cursors_xdg[] = {
+	NULL,
+	"default",
+	"grab",
+	"nw-resize",
+	"n-resize",
+	"ne-resize",
+	"e-resize",
+	"se-resize",
+	"s-resize",
+	"sw-resize",
+	"w-resize"
 };
 
+/* XCursor fallbacks */
+static const char *cursors_x11[] = {
+	NULL,
+	"left_ptr",
+	"grabbing",
+	"top_left_corner",
+	"top_side",
+	"top_right_corner",
+	"right_side",
+	"bottom_right_corner",
+	"bottom_side",
+	"bottom_left_corner",
+	"left_side"
+};
+
+#define ARRAY_SIZE(array) (sizeof(array) / sizeof((array)[0]))
+static_assert(
+	ARRAY_SIZE(cursors_xdg) == LAB_CURSOR_COUNT,
+	"XDG cursor names are out ot sync");
+static_assert(
+	ARRAY_SIZE(cursors_x11) == LAB_CURSOR_COUNT,
+	"X11 cursor names are out ot sync");
+#undef ARRAY_SIZE
+
+static enum lab_cursors
+cursor_get_from_ssd(enum ssd_part_type view_area)
+{
+	switch (view_area) {
+	case LAB_SSD_PART_CORNER_TOP_LEFT:
+		return LAB_CURSOR_RESIZE_NW;
+	case LAB_SSD_PART_TOP:
+		return LAB_CURSOR_RESIZE_N;
+	case LAB_SSD_PART_CORNER_TOP_RIGHT:
+		return LAB_CURSOR_RESIZE_NE;
+	case LAB_SSD_PART_RIGHT:
+		return LAB_CURSOR_RESIZE_E;
+	case LAB_SSD_PART_CORNER_BOTTOM_RIGHT:
+		return LAB_CURSOR_RESIZE_SE;
+	case LAB_SSD_PART_BOTTOM:
+		return LAB_CURSOR_RESIZE_S;
+	case LAB_SSD_PART_CORNER_BOTTOM_LEFT:
+		return LAB_CURSOR_RESIZE_SW;
+	case LAB_SSD_PART_LEFT:
+		return LAB_CURSOR_RESIZE_W;
+	default:
+		return LAB_CURSOR_DEFAULT;
+	}
+}
+
+enum lab_cursors
+cursor_get_from_edge(enum wlr_edges edge)
+{
+	/* Cast to uint32 because TOP | LEFT is not actually part of the enum */
+	switch ((uint32_t)edge) {
+	case WLR_EDGE_TOP | WLR_EDGE_LEFT:
+		return LAB_CURSOR_RESIZE_NW;
+	case WLR_EDGE_TOP:
+		return LAB_CURSOR_RESIZE_N;
+	case WLR_EDGE_TOP | WLR_EDGE_RIGHT:
+		return LAB_CURSOR_RESIZE_NE;
+	case WLR_EDGE_RIGHT:
+		return LAB_CURSOR_RESIZE_E;
+	case WLR_EDGE_BOTTOM | WLR_EDGE_RIGHT:
+		return LAB_CURSOR_RESIZE_SE;
+	case WLR_EDGE_BOTTOM:
+		return LAB_CURSOR_RESIZE_S;
+	case WLR_EDGE_BOTTOM | WLR_EDGE_LEFT:
+		return LAB_CURSOR_RESIZE_SW;
+	case WLR_EDGE_LEFT:
+		return LAB_CURSOR_RESIZE_W;
+	default:
+		wlr_log(WLR_ERROR,
+			"Failed to resolve wlroots edge %u to cursor name", edge);
```

Comment:
(Trivial) I'd be inclined to omit the log message here, in such a "utility" style function as this.

Reply:
I think we should never receive some invalid edge combination here. If we do, its a bug so I added an `assert(false)` after the error. I am open to change it to just return the default cursor and remove the log though.